src/coreclr/nativeaot/System.Private.CoreLib/src/Internal/Runtime/CompilerHelpers/MathHelpers.cs

```diff
@@ -2,381 +2,329 @@
+            // We need to compare with the very next double to two63. 0x402 is epsilon to get us there.
+            if (value is > -two63 - 0x402 and < two63)
+            {
+                long ret = (long)value;
+                // since no overflow can occur, the value always has to be within 1
+                Debug.Assert(value - 1.0 <= ret);
+                Debug.Assert(ret <= value + 1.0);
+                return ret;
+            }
+
+            ThrowHelper.ThrowOverflowException();
+            return 0;
         }
 
-        // Helper to get high 32-bit of 64-bit int
-        [MethodImpl(MethodImplOptions.AggressiveInlining)]
-        private static uint Hi32Bits(ulong a)
+        [RuntimeExport("Dbl2ULngOvf")]
+        public static ulong Dbl2ULngOvf(double value)
         {
-            return (uint)(a >> 32);
+            const double two64 = UIntMaxValueOffset * UIntMaxValueOffset;
+            // Note that this expression also works properly for val = NaN case
+            if (value is > -1.0 and < two64)
+            {
+                ulong ret = (ulong)value;
+                // since no overflow can occur, the value always has to be within 1
+                Debug.Assert(value - 1.0 <= ret);
+                Debug.Assert(ret <= value + 1.0);
+                return ret;
+            }
+
+            ThrowHelper.ThrowOverflowException();
+            return 0;
         }
 
+#if !TARGET_64BIT
+        //
+        // 64-bit checked multiplication for 32-bit platforms
+        //
+
+        private const string RuntimeLibrary = "*";
+
         [RuntimeExport("LMulOvf")]
-        public static long LMulOvf(long i, long j)
+        public static long LMulOvf(long left, long right)
         {
-            long ret;
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            static uint High32Bits(ulong a)
+            {
+                return (uint)(a >> 32);
+            }
+
+#if DEBUG
+            long result = left * right;
+#endif
 
             // Remember the sign of the result
-            int sign = (int)(Hi32Bits(i) ^ Hi32Bits(j));
+            int sign = (int)(High32Bits((ulong)left) ^ High32Bits((ulong)right));
 
             // Convert to unsigned multiplication
-            if (i < 0) i = -i;
-            if (j < 0) j = -j;
+            if (left < 0)
+                left = -left;
+            if (right < 0)
+                right = -right;
 
             // Get the upper 32 bits of the numbers
-            uint val1High = Hi32Bits(i);
-            uint val2High = Hi32Bits(j);
+            uint val1High = High32Bits((ulong)left);
+            uint val2High = High32Bits((ulong)right);
 
             ulong valMid;
 
             if (val1High == 0)
             {
                 // Compute the 'middle' bits of the long multiplication
-                valMid = Mul32x32To64(val2High, (uint)i);
+                valMid = Math.BigMul(val2High, (uint)left);
```

Comment:
Switching this to Math.BigMul seems to be a deoptimization when compiling for VEX-enabled baseline. The Bmi2 special path in Math.BigMul is slower than a regular multiplication that's here today:

My usual lame microbenchmark:
```
using System;
using System.Diagnostics;
using System.Runtime.CompilerServices;

for (; ; )
{
    var sw = new Stopwatch();
    sw.Start();
    Work(2_000_000_000);
    Console.WriteLine(sw.ElapsedMilliseconds);
}

[MethodImpl(MethodImplOptions.NoInlining)]
long Work(long x)
{
    long sum = 0;
    for (long i = 1; i < x; i += 10)
        sum += checked(i * i);
    return sum;
}
```

Baseline: ~680ms
PR w/ --instruction-set:bmi2: ~800ms

Reply:
cc @tannergooding do we want to remove the BMI2 path there in this case?

Reply:
It actually seems like `SkipLocalsInit` is making the JIT confused in the BMI2 path:
```asm
G_M56079_IG01:        ; bbWeight=1, gcrefRegs=00000000 {}, byrefRegs=00000000 {}, byref, nogc <-- Prolog IG
       56           push     esi
       83EC08       sub      esp, 8
						;; size=4 bbWeight=1 PerfScore 1.25
G_M56079_IG02:        ; bbWeight=1, gcrefRegs=00000000 {}, byrefRegs=00000000 {}, byref
                    ; INLRT @ 0x003[E-]
       8B0424       mov      eax, dword ptr [esp]
       23D0         and      edx, eax
       89542404     mov      dword ptr [esp+0x04], edx
       8B742404     mov      esi, dword ptr [esp+0x04]
       8BD1         mov      edx, ecx
       C4E26BF6D6   mulx     edx, edx, esi
       83C800       or       eax, 0
       83CA00       or       edx, 0
						;; size=26 bbWeight=1 PerfScore 7.00
G_M56079_IG03:        ; bbWeight=1, epilog, nogc, extend
       83C408       add      esp, 8
       5E           pop      esi
       C3           ret      
						;; size=5 bbWeight=1 PerfScore 1.75
```

Without that:
```asm
G_M56079_IG01:        ; bbWeight=1, gcrefRegs=00000000 {}, byrefRegs=00000000 {}, byref, nogc <-- Prolog IG
						;; size=0 bbWeight=1 PerfScore 0.00
G_M56079_IG02:        ; bbWeight=1, gcrefRegs=00000000 {}, byrefRegs=00000000 {}, byref
                    ; INLRT @ 0x003[E-]
       33C0         xor      eax, eax
       8BD1         mov      edx, ecx
       C4E26BF6D0   mulx     edx, edx, eax
						;; size=9 bbWeight=1 PerfScore 3.50
G_M56079_IG03:        ; bbWeight=1, gcrefRegs=00000000 {}, byrefRegs=00000000 {}, byref, epilog, nogc
       C3           ret      
						;; size=1 bbWeight=1 PerfScore 1.00
```

Reply:
FWIW, I think mulx should be an optimization in the JIT that automatically kicks in for all multiplications that fit the pattern. It does not make sense to make everybody out there to tweak their 32-bit*32-bit->64-bit multiplication to deal with the BigMul helper.